### ISSUE TITLE
-Wsizeof-pointer-memaccess doesn't catch memset( itemInfo, 0, sizeof( &itemInfo ) )

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#12661]
|  |  |
| --- | --- |
| Bugzilla Link | [12289](https://llvm.org/bz12289) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
The following program does not produce a warning when compiling with -Wsizeof-pointer-memaccess

```cpp
#include <cstring>

struct Foo {};

int main() {
    Foo p, *itemInfo = &p;
    
    std::memset( itemInfo, 0, sizeof( &itemInfo ) );

}
```

If the memset line were slightly different, a warning would be generated:

```
std::memset( itemInfo, 0, sizeof( itemInfo ) );
```

This code was tested due to the following article (Example 1):

http://www.viva64.com/en/a/0079/#ID0EYKAG

**Comment by chfast** [comment]
The warning is limited to some specific cases and cannot handle arbitrary types. It is triggered in case of passing `Foo*, sizeof(Foo*)` (second case) but not for `Foo*, sizeof(Foo**)` (first case).

GCC diagnostics work the same way.

https://godbolt.org/z/hPnsecvYK

**Comment by Endilll** [comment]
I believe this in an enhancement that haven't been implemented yet.
